sdk/identity/azure-identity/src/azure_cli_credential.cpp

```diff
@@ -0,0 +1,654 @@
+  {
+    // Since the strings that argv is pointing at do need to be non-const,
+    // and also because each commnd line argument needs to be a separate 0-terminated string,
+    // We do form all their values in the m_argvValues.
+
+    // Since we append m_argvValues as we go, at one point after insertion it may reallocate the
+    // buffer to a different address in memory. For that reason, we can't grab addresses before we
+    // are done forming m_argvValues contents - so until that we record indices where each string
+    // start - in argvValuePositions.
+    {
+      std::vector<decltype(m_argvValues)::size_type> argvValuePositions;
+
+      // First string is the path to executable, and not the actual first argument.
+      {
+        std::string const Shell = "/bin/sh";
+        EnsureShellExists(Shell);
+        AppendToArgvValues(m_argvValues, argvValuePositions, Shell);
+      }
+
+      // Second argument is the shell switch that tells the command line shell to execute a command
+      AppendToArgvValues(m_argvValues, argvValuePositions, "-c");
+
+      // Third value is the command that needs to be executed.
+      AppendToArgvValues(m_argvValues, argvValuePositions, command);
+
+      // We are done appending to m_argvValues, so it is now safe to grab addresses to the elements
+      // in it.
+      for (auto const pos : argvValuePositions)
+      {
+        m_argv.push_back(m_argvValues.data() + pos);
+      }
+    }
+
+    // argv last element needs to be NULL pointer.
+    m_argv.push_back(NULL);
+  }
+
+  // Form the 'envp' array:
+  // * An array of pointers to non-const C strings (0-terminated).
+  // * Strings are in form key=value (PATH uses ':' as separator)
+  // * Last element is NULL pointer.
+  // * First element (at index 0) is path to a program.
+  {
+    auto const actualPathVarValue = Environment::GetVariable("PATH");
+    auto const processPathVarStatement = std::string("PATH=") + actualPathVarValue
+        + (actualPathVarValue.empty() ? "" : ":") + "/usr/bin:/usr/local/bin";
+
+    m_envpValues.insert(
+        m_envpValues.end(), processPathVarStatement.begin(), processPathVarStatement.end());
+
+    m_envpValues.push_back('\0');
+
+    // We should only grab m_envpValues.data() as we're done appending to it, because appends may
+    // reallocate the buffer to a different memory location.
+    m_envp.push_back(m_envpValues.data());
+    m_envp.push_back(NULL);
+  }
+
+  // Set up pipe communication for the process.
+  posix_spawn_file_actions_init(&m_actions);
+  posix_spawn_file_actions_addclose(&m_actions, outputPipe.m_fd[0]);
+  posix_spawn_file_actions_adddup2(&m_actions, outputPipe.m_fd[1], 1);
+  posix_spawn_file_actions_addclose(&m_actions, outputPipe.m_fd[1]);
+
+  {
+    auto const spawnResult
+        = posix_spawn(&m_pid, m_argv[0], &m_actions, NULL, m_argv.data(), m_envp.data());
+
+    // LCOV_EXCL_START
+    if (spawnResult != 0)
+    {
+      m_pid = -1;
+      this->~ShellProcess();
```

Comment:
Generally it's bad form to explicitly call the destructor of a class - normally you have a "close" helper function which is called by the destructor and the call site where you're tempted to call the destructor.

Reply:
I know, but it is internal namespace, local class, I did not want to add more lines to complicate, and was relying on the fact that both functions are close and we know what we're doing, because both places can fit on a single screen. But I can create a `Destruct()` function that gets called from both here and from constructor.

Reply:
This is one of those "just because the compiler allows it doesn't mean it's right" situations. See also [this StackOverflow topic](https://stackoverflow.com/questions/14187006/is-calling-destructor-manually-always-a-sign-of-bad-design#:~:text=You%20call%20the%20dtor%20explicitly%20every%20time%20you,is%20managed%20independently%20from%20object%20construction%20%2F%20destruction.).

There are other reasons to manually call the destructor (for instance, in an RAII pointer, the operator= can often call the destructor to invoke destruct sematics without releasing the underlying storage). But I don't see them here.